File: reference/macos-swiftui/AirX/Util/AccountUtils.swift
//
//  AccountUtil.swift
//  AnyDropmac
//
//  Created by Hatsune Miku on 2023-06-14.
//

import Foundation
import SwiftUI

class AccountUtils {
    private static var subscribers = Dictionary<String, (Bool) -> Void>()

    public static func subscribeToAutomaticLoginResult(id: String, handler: @escaping (Bool) -> Void) {
        subscribers[id] = handler
    }
    
    public static func clearSavedUserInfoAndSignOut() {
        Defaults.delete(.savedCredential)
        Defaults.delete(.loggedInUid)
        Defaults.delete(.savedCredentialType)
    }
    
    public static func notifySubscribers(didLoginSuccess: Bool) {
        for subscriber in subscribers.values {
            subscriber(didLoginSuccess)
        }
    }
    
    /**
     * Return: true if successfully logged in, otherwise, false.
     */
    public static func tryLoginWithSavedToken() {
        GlobalState.shared.isSignedIn = false
        print("Trying automatic login...")

        // Check credentials!
        guard Defaults.credentialType() == .anydropToken else {
            print("Failed: incorrect credential type")
            notifySubscribers(didLoginSuccess: false)
            return
        }
        
        let token = Defaults.string(.savedCredential, def: "")
        guard !token.isEmpty else {
            print("Failed: empty token")
            notifySubscribers(didLoginSuccess: false)
            return
        }
        
        let uid = Defaults.string(.savedUsername, def: "")
        guard !uid.isEmpty else {
            print("Failed: empty uid")
            notifySubscribers(didLoginSuccess: false)
            return
        }
        
        
        do {
            try AnyDropCloud.renew(uid: uid) { response in
                guard response.success else {
                    print("Failed: renew failed: \(response.message)")
                    notifySubscribers(didLoginSuccess: false)
                    return
                }
                
                // Almost there!
                print("Token renewal success.")
                GlobalState.shared.isSignedIn = true
                Defaults.write(.loggedInUid, value: uid)
                Defaults.write(.savedCredential, value: response.token)
                
                /// Register with the backend
                if !WebSocketService.shared.initialize() {
                    print("Failed to register with backend.")
                }
            }
        }
        catch {
            notifySubscribers(didLoginSuccess: false)
            return
        }
        notifySubscribers(didLoginSuccess: true)
    }
    
    public static func blockUser(peer: Peer) {
        // TODO: 
    }
}

File: reference/windows-winui/AirX/Utils/AccountUtils.cs
using AnyDrop.Bridge;
using AnyDrop.Extension;
using AnyDrop.Model;
using AnyDrop.Services;
using AnyDrop.View;
using AnyDrop.ViewModel;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AnyDrop.Util
{
    public class AccountUtils
    {
        private static HashSet<string> _blockList = SettingsUtils.ReadBlockList();

        public static void ClearSavedUserInfoAndSignOut()
        {
            SettingsUtils.Delete(Keys.SavedCredential);
            SettingsUtils.Delete(Keys.LoggedInUid);
            GlobalViewModel.Instance.IsSignedIn = false;
            GlobalViewModel.Instance.LoggingInUid = "";
            GlobalViewModel.Instance.LoggingGreetingsName = "AnyDrop User";
        }

        public static bool IsInBlockList(string ipAddress)
        {
            if (!Peer.TryParse(ipAddress, out Peer peer))
            {
                return false;
            }
            return _blockList.Contains(peer.IpAddress);
        }

        public static void AddToBlockList(string ipAddress)
        {
            if (!Peer.TryParse(ipAddress, out Peer peer))
            {
                return;
            }
            _blockList.Add(peer.IpAddress);
            SettingsUtils.WriteBlockList(_blockList);
        }

        public static async Task<bool> SendGreetingsAsync()
        {
            AnyDropCloud.GreetingsResponse greetingsResponse;
            try
            {
                var uid = SettingsUtils.String(Keys.SavedUid, "");
                greetingsResponse = await AnyDropCloud.GreetingsAsync();
                if (greetingsResponse.success)
                {
                    GlobalViewModel.Instance.LoggingGreetingsName = greetingsResponse.name;
                    return true;
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine($"Failed: greetings failed: {e.Message}");
            }

            return false;
        }

        /**
         * Return: true if successfully logged in, otherwise, false.
         */
        public static async Task<bool> TryLoginWithSavedTokenAsync()
        {
            if (!SettingsUtils.Bool(Keys.ShouldAutoSignIn, false))
            {
                return false;
            }

            GlobalViewModel.Instance.IsSignedIn = false;
            Debug.WriteLine("Trying automatic login...");

            // Check credentials!
            if (SettingsUtils.ReadCredentialType() != CredentialType.AnyDropToken)
            {
                Debug.WriteLine("Failed: incorrect credential type");
                return false;
            }

            string token = SettingsUtils.String(Keys.SavedCredential, "");
            if (string.IsNullOrEmpty(token))
            {
                Debug.WriteLine("Failed: empty token");
                return false;
            }

            string uid = SettingsUtils.String(Keys.SavedUid, "");
            if (string.IsNullOrEmpty(uid))
            {
                Debug.WriteLine("Failed: empty uid");
                return false;
            }

            AnyDropCloud.RenewResponse renewResponse;
            try
            {
                renewResponse = await AnyDropCloud.RenewAsync(uid);
                if (renewResponse == null || !renewResponse.success)
                {
                    Debug.WriteLine($"Failed: renew failed.");
                    return false;
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine($"Failed: renew failed: {e.Message}");
                return false;
            }

            await SendGreetingsAsync();

            Debug.WriteLine("Token renewal success.");

            // Update state.
            GlobalViewModel.Instance.LoggingInUid = uid;
            GlobalViewModel.Instance.IsSignedIn = true;

            // Update storage.
            SettingsUtils.Write(Keys.LoggedInUid, uid);
            SettingsUtils.Write(Keys.SavedCredential, renewResponse.token);

            // Initialize websocket.
            WebSocketService.Instance.InitializeAsync().ContinueWith(task =>
            {
                Debug.WriteLine(
                    task.Result
                        ? "Successfully initialized websocket."
                        : "Failed to initialize websocket."
                );
            }, TaskScheduler.Default).FireAndForget();

            return true;
        }

        public static void UserToggleSignInOut()
        {
            if (GlobalViewModel.Instance.IsSignedIn)
            {
                GlobalViewModel.Instance.IsSignedIn = false;
                ClearSavedUserInfoAndSignOut();
                return;
            }

            var window = new LoginWindow();
            window.Activate();
        }
    }
}
